buildtest/buildsystem/base.py

```diff
@@ -78,14 +79,15 @@ def load(self, buildspec):
 
            Parameters:
 
-           buildspec: the pull path to the configuration file, must exist.
+           :param buildspec: the pull path to the configuration file, must exist.
+           :type buildspec: str, required
         """
         self.buildspec = os.path.abspath(buildspec)
 
-        if not os.path.exists(self.buildspec):
+        if not is_file(self.buildspec):
             sys.exit("Buildspec File: %s does not exist." % self.buildspec)
 
-        elif os.path.isdir(self.buildspec):
+        elif is_dir(self.buildspec):
```

Comment:
This change doesn't make sense - if you just checked with is_file this statement will never be reached. The original check was intended to first check if the path exists, and then parse if it's a directory or not (hitting the second statement). Now this is unreachable, unless there is some niche case of a symlink not being identified as a file and then hitting the next one.

Reply:
Yea I see your point, the elif wont be reached i was testing this locally just for my sanity and hope we can adopt these methods as much as we can. 

is_file works with shell expansion and works by running ``os.path.isfile`` underneath
```
>>> is_file("$HOME/.buildtest/settings.yml")
True
>>> is_file("$HOME/.buildtest/")
False
```

If a directory is specified ``is_file`` will return ``False`` and the ``if not is_file`` will cause statment to be ``True`` and never reach the elif.

The thing is os.path.exists wouldn't work for some edge cases such as 

```
>>> os.path.exists("$HOME/.buildtest/settings.yml")
False
>>> is_file("$HOME/.buildtest/settings.yml")
True
>>> os.path.exists("$HOME/.buildtest/")
False
>>> is_dir("$HOME/.buildtest/")
True
```
We also account for tilde ``~`` if passed as input but os.path.exists won't catch this

```
>>> is_file("~/.buildtest/settings.yml")
True
>>> os.path.exists("~/.buildtest/settings.yml")
False
```
Symlinks work with os.path.exists so that's not an issue. It's only for discovery and exclusion we get the real path in-order for exclusion to work correctly.
The reason why i changed os.path.exists to ``is_file`` or ``is_dir`` depending on context of file is to account for tilde or shell expansion
```
>>> os.path.islink("/u/users/ssi29/gpfs")
True
>>> resolve_path("/u/users/ssi29/gpfs")
'/mxg-hpc/users/ssi29'
```

Note that ``is_file`` and ``is_dir`` are calling ``resolve_path`` before the check.

Reply:
So under what conditions would the second elif trigger?

Reply:
Yea i fixed this now in latest commit and let me demo this 

If input is a directory

```
>>> spec = BuildspecParser("$PWD/examples/benchmarks/")
Detected /mxg-hpc/users/ssi29/buildtest-framework/examples/benchmarks is a directory, please provide a file path (not a directory path) to BuildspecParser.
```

If its invalid file

```
>>> spec = BuildspecParser("xyz!@#")
Can't process input: xyz!@#
```

If its valid file we proceed as normal

```
>>> spec = BuildspecParser("$PWD/examples/benchmarks/stream.yml")
spec.get(           spec.keys(          spec.logger         spec.metadata
>>> spec.recipe
{'version': '0.0.1', 'stream_uniprocess_c': {'type': 'script', 'description': 'STREAM Microbenchmark C test on uniprocessor', 'run': 'wget https://www.cs.virginia.edu/stream/FTP/Code/stream.c\ngcc -O2 stream.c -o stream\n./stream\n'}, 'stream_openmp_c': {'type': 'script', 'description': 'STREAM Microbenchmark C Test with OpenMP', 'run': 'wget https://www.cs.virginia.edu/stream/FTP/Code/stream.c\ngcc -fopenmp -D_OPENMP -O2 stream.c -o stream\n\nexport OMP_NUM_THREAD=2\n./stream\n\nexport OMP_NUM_THREAD=4\n./stream\n\nexport OMP_NUM_THREAD=8\n./stream\n\nexport OMP_NUM_THREAD=16\n./stream\n'}}

```